File: sdk/python/polos/utils/tracing.py
from opentelemetry.trace import NonRecordingSpan, set_span_in_context


def get_parent_span_context_from_execution_context(exec_context):
    parent_context = None
    if exec_context:
        parent_span_context = exec_context.get("_otel_span_context")
        if parent_span_context:
            # Convert SpanContext to Context by wrapping in NonRecordingSpan
            parent_span = NonRecordingSpan(parent_span_context)
            parent_context = set_span_in_context(parent_span)

    return parent_context


def get_span_context_from_execution_context(exec_context):
    if exec_context:
        return exec_context.get("_otel_span_context")
    return None


def set_span_context_in_execution_context(exec_context, span_context):
    if exec_context:
        exec_context["_otel_span_context"] = span_context
        if span_context is not None:
            exec_context["_otel_trace_id"] = format(span_context.trace_id, "032x")
            exec_context["_otel_span_id"] = format(span_context.span_id, "016x")

File: sdk/typescript/src/utils/tracing.ts
/**
 * Span context utilities for ExecutionContextData.
 *
 * Matches Python's utils/tracing.py. Manages the integration between
 * Polos execution context (AsyncLocalStorage) and OTel span contexts.
 */

import { type Context, type SpanContext, trace, context as otelContext } from '@opentelemetry/api';
import type { ExecutionContextData } from '../runtime/execution-context.js';

/**
 * Get parent OTel context from execution context.
 * Wraps the stored SpanContext in a NonRecordingSpan and sets it in an OTel Context,
 * suitable for passing as `context` to `tracer.startSpan(name, opts, context)`.
 */
export function getParentSpanContextFromExecutionContext(
  execContext: ExecutionContextData | undefined
): Context | undefined {
  if (!execContext) return undefined;

  const parentSpanContext = execContext._otelSpanContext as SpanContext | undefined;
  if (!parentSpanContext) return undefined;

  // Wrap SpanContext in NonRecordingSpan → set in OTel Context
  const parentSpan = trace.wrapSpanContext(parentSpanContext);
  return trace.setSpan(otelContext.active(), parentSpan);
}

/**
 * Get raw span context from execution context.
 */
export function getSpanContextFromExecutionContext(
  execContext: ExecutionContextData | undefined
): SpanContext | undefined {
  if (!execContext) return undefined;
  return execContext._otelSpanContext as SpanContext | undefined;
}

/**
 * Store span context in execution context along with formatted trace ID and span ID.
 */
export function setSpanContextInExecutionContext(
  execContext: ExecutionContextData | undefined,
  spanContext: SpanContext | undefined
): void {
  if (!execContext) return;

  execContext._otelSpanContext = spanContext;
  if (spanContext != null) {
    execContext._otelTraceId = spanContext.traceId;
    execContext._otelSpanId = spanContext.spanId;
  }
}
